### ISSUE TITLE
SFINAE friendly signatures

### ISSUE THREAD

**Issue by JohelEGP** [nholthaus/units#114]
Take a look at the following code:
```C++
using namespace units::literals;
static_assert(!std::is_convertible_v<decltype(0_m), decltype(0_m * 0_m)>); // [1]
units::length::meter_t m{0_m * 0_m}; // [2]
```
[1] triggers the assert although [2] is an error. That's because the signatures that are used to check the type trait in [1] are valid while the error in [2] is caused by an `static_assert` in the body of the `convert` function.

The fix to get this to work would be to move any late checking done in `convert` to an early check in the signatures of the functions that directly or indirectly call it.

The lack of SFINAE friendliness in the library is causing the concepts I linked to earlier not to work as expected.

**Comment by johelegp** [comment]
Thanks for the warning. I'm well aware of it.

> ... That's part of the big refactor that needs to be done for 3.0 ...

Hopefully that means that during the refactor the checks done by `convert` will be moved by to the signatures that use it. That'd be the most general and clean solution, wouldn't require playing favorites which would just add baggage to this library's code, nor require general solutions, like the concepts, to know about `units::traits`.

> ... I can prioritize your favorites ...

Thanks for the offer, but the problem is non-blocking.

**Comment by johelegp** [comment]
When implementing this change, you should consider changing the primary templates of unit type traits to not define the type aliases to `void`, and do like the standard type traits and define no type aliases.

**Comment by johelegp** [comment]
To implement this, what before might have been a compile error with some nice static assertion messages, would become a "novel" of errors, as said in [the video](https://youtu.be/PFdWqa68LmA) linked in the article [How to Make SFINAE Pretty – Part 1: What SFINAE Brings to Code](https://www.fluentcpp.com/2018/05/15/make-sfinae-pretty-1-what-value-sfinae-brings-to-code/) from FluentC++. Said video (and presumably the follow-up of the linked article) show how we might mitigate this. I'm already [using that technique](https://github.com/nholthaus/units/pull/126/commits/e48ed1721173a5440fa0b52d69dd1b5793763c59#diff-cc75003c399d878c360873c2c8c783d6R1549) in #126, which consists on factoring uses of `std::enable_if_t` with more appropriate names. Another use might be to name the boolean expressions, as modern compiler give special treatment to the first argument of `std::enable_if_t` to make errors messages clearer.

Even then, the units, being type aliases rather than actual types, will continue to be spelled out in full in the error messages. It has been suggested before to make them different types.

**Comment by nholthaus** [comment]
#150 or one of the other PRs made us a lot more sfinae friendly right? If not I don't think we'll do this for 3.0.0

**Comment by johelegp** [comment]
Not really. It only slightly improved existing use of SFINAE. Many signatures still need them, like the in-class unit ordering operators, which should check for `is_same_dimension`.

**Comment by nholthaus** [comment]
Resolved by the 3.x concepts rewrite: the dimension check moved out of `convert`'s body into its signature as a `requires` constraint (`is_same_dimension_conversion_factor_v` / `same_dimension`), and `convert` no longer contains a body `static_assert`, so a dimensionally-incompatible conversion is SFINAE-friendly (fails substitution rather than hard-erroring). See `convert` and the unit converting constructor in `include/units/core.h`.
